File: loreshifter/lib/core/models/game.dart
import 'package:loreshifter/core/models/player.dart';
import 'package:loreshifter/core/models/world.dart';

enum GameStatus { waiting, playing, finished, archived }

class Game {
  final int id;
  final String code;
  final bool public;
  final String name;
  final World world;
  final int hostId;
  final List<Player> players;
  final DateTime createdAt;
  final int maxPlayers;
  final GameStatus status;

  Game({
    required this.id,
    required this.code,
    required this.public,
    required this.name,
    required this.world,
    required this.hostId,
    required this.players,
    required this.createdAt,
    required this.maxPlayers,
    required this.status,
  });

  factory Game.fromJson(Map<String, dynamic> json) {
    return Game(
      id: json['id'] as int,
      code: json['code'] as String,
      public: json['public'] as bool,
      name: json['name'] as String,
      world: World.fromJson(json['world']),
      hostId: json['hostId'] as int,
      players: (json['players'] as List)
          .map((e) => Player.fromJson(e as Map<String, dynamic>))
          .toList(),
      createdAt: DateTime.parse(json['createdAt']),
      maxPlayers: json['maxPlayers'] as int,
      status: _parseGameStatus(json['status'] as String),
    );
  }

  static GameStatus _parseGameStatus(String status) {
    switch (status) {
      case 'waiting':
        return GameStatus.waiting;
      case 'playing':
        return GameStatus.playing;
      case 'finished':
        return GameStatus.finished;
      case 'archived':
        return GameStatus.archived;
      default:
        return GameStatus.waiting;
    }
  }

  Map<String, dynamic> toJson() {
    return {
      'id': id,
      'code': code,
      'public': public,
      'name': name,
      'world': world.toJson(),
      'hostId': hostId,
      'players': players.map((e) => e.toJson()).toList(),
      'createdAt': createdAt.toIso8601String(),
      'maxPlayers': maxPlayers,
      'status': status.toString().split('.').last,
    };
  }
}

File: backend/lstypes/game.py
import dataclasses
import datetime
import enum

from lstypes.chat import ChatSegmentOut
from lstypes.player import PlayerOut
from lstypes.utils import PgEnum
from lstypes.world import ShortWorldOut


class GameStatus(enum.Enum, metaclass=PgEnum):
    __pg_enum_name__ = "game_status"
    WAITING = "waiting"
    PLAYING = "playing"
    FINISHED = "finished"
    ARCHIVED = "archived"


@dataclasses.dataclass
class GameOut:
    id: int
    code: str
    public: bool
    name: str
    world: ShortWorldOut
    host_id: int
    players: list[PlayerOut]
    created_at: datetime.datetime
    max_players: int
    status: GameStatus


@dataclasses.dataclass
class StateOut:
    game: GameOut
    status: GameStatus
    character_creation_chat: ChatSegmentOut | None = None
    game_chat: ChatSegmentOut | None = None
    player_chats: list[ChatSegmentOut] = dataclasses.field(default_factory=list)
    advice_chats: list[ChatSegmentOut] = dataclasses.field(default_factory=list)
